### DISCUSSION TITLE
Understanding the forecasting and alert functionality

### DISCUSSION THREAD

**Discussion by js1972** [SAP-samples/btp-resource-consumption-monitor#10] [Q&A]
Awesome app. Just got it up and running. I notice the 3 hourly job is continually sending me alerts due to the forecast values being 200 - 300% over current actual values. I just have all the default values at this stage for the forecasts. Why would it be forecasting so high. Am I supposed to go through each service and metric and individually set these? Even if I change this to Linear forecasting the percentage forecasts are very high - over 400%.

Also - if I set all forecasts to EXCLUDED - does that now mean I will only receive alerts for actual cost increases?

What I am tying to work out is how to get an alert only when there is a cost blowout. I.e. someone turns on an additional integration suite tenant and triggers a $6k charge. What settings would be necessary to get an alert for that?

**Comment by willempardaens** [comment]
Hi Jason, thank you for your feedback and the question. Some of the explanation of the forecast configuration is available here https://github.com/SAP-samples/btp-resource-consumption-monitor#forecasting-configuration, but to be more precise to your query:

Keeping default values: indeed, it would be advised to configure each service (and each of their commercial metrics) to set the correct forecast setting. E.g. for SAP Integration Suite tenants, the `excluded` value makes most sense, as that is a measure that has no time aspect ito forecasting: if you start the month with 3 tenants, you will probably end the month also with 3 tenants, hence, no increase to be forecasted.

This indeed means that for a metric configured as `excluded`, the forecast will be equal to the latest actual.

For the scenario with the cost blowout I think that is a gap in the tool as of now: the tool focusses on spikes in forecasts, not spikes in actuals for services that are not forecasted, as this would typically be the result of manual actions. It certainly makes sense to add that capability by, e.g., giving you access to a 'usage-delta' variable. I'll explore how to easily enable this. Thanks for describing this scenario.

**Reply by js1972** [reply]
Thankyou @willemi069808 - with regards to the No Foreacast example you gave with integration suite... If a customer has 3 tenants you do not expect that to change.
But if a 4th tenant was added - would it be seen in the next alert email. The number of tenants has changed so it would be good for an admin to know this but the name of this forecast method is "no forecast" and the alerting seems to be only based on forecasts which would suggest this would not get added to the next alert??

Using another example - if there were 1000 active users in work zone and you set the forecast method to No Forecast. Then the following month there was 5000 active users - would that show on the alert email?

The usage delta variable is a great idea and is a much needed case with customers. You would not know the number of times someone at a customer turns on services without management knowing till they get the bill shock!

**Reply by willempardaens** [reply]
Hi @js1972 , I added delta calculations. Kindly test the branch in #11 and let me know what you think.

**Reply by js1972** [reply]
Awesome. Have just pulled down the PR and deployed to my cf org. But... seems to have removed the app from the HTML5 Applications menu in the subaccount and launching from work zone doesn't work. Any ideas? Have cleared caches and also manually refreshed the WZ content providers. Didn't help.

Maybe I need to do a full undeploy and the re-deploy?

Actually I can see this error:

<img width="496" alt="image" src="https://github.com/SAP-samples/btp-resource-consumption-monitor/assets/1317161/f24c5b0e-2e0f-49a5-9a55-43d28aa760f3">

There are no errors in the cf deployment log.

**Reply by js1972** [reply]
fyi. a redeployment fixed that strange btprc-dest error. The app still does not show in the subaccount html5 apps menu nor. does it run from work zone.

Chrome logs:
```
Uncaught (in promise) TypeError: Cannot create property 'url' on string '<!DOCTYPE HTML>
<html>
<head>
    <meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>UI5 Fiori Application Runtime</title>

    <link rel="stylesheet" crossorigin="anonymous" href="https&#x3a;&#x2f;&#x2f;sapui5.hana.ondemand.com&#x2f;1.123.1&#x2f;resources&#x2f;sap&#x2f;ui&#x2f;core&#x2f;themes&#x2f;sap_horizon&#x2f;library.css&#x3f;version&#x3d;1.123.1&amp;sap-ui-dist-version&#x3d;1.123.1" id="sap-ui-theme-sap.ui.core" data-sap-ui-ready="true"/>
    <link rel="stylesheet" crossorigin="anonymous" href="https&#x3a;&#x2f;&#x2f;sapui5.hana.ondemand.com&#x2f;1.123.1&#x2f;resources&#x2f;sap&#x2f;m&#x2f;themes&#x2f;sap_horizon&#x2f;library.css&#x3f;version&#x3d;1.123.1&amp;sap-ui-dist-version&#x3d;1.123.1" id="sap-ui-theme-sap.m" data-sap-ui-ready="true"/>

    <meta name="sap-allowedThemeOrigins" content='https://teams.microsoft.com,https://js-4f5krbai.cpp.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.launchpad.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.launchpad.v1.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.ap10.workzone.cloud.sap,https://js-4f5krbai.workzone.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.workzonehr.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.mobile.workzone.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.mobile.workzonehr.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.cppng.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.launchpadng.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.workzoneng.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.workzonehrng.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.ap10.start.cloud.sap,https://js-4f5krbai.mobile.workzoneng.cfapps.ap10.hana.ondemand.com,https://js-4f5krbai.mobile.workzonehrng.cfapps.ap10.hana.ondemand.com,https://sapui5.hana.ondemand.com'/>
    <meta name="sap.flp.cf.Config" content='&#x7b;&quot;fdcId&quot;&#x3a;&quot;4e8bdc0e_8fac_437b_b2ed_d7df8327d396&quot;,&quot;subaccountId&quot;&#x3a;&quot;4e8bdc0e-8fac-437b-b2ed-d7df8327d396&quot;&#x7d;'/>
    <meta name="sap.ushellConfig.ui5appruntime" content='{
            "flpPlatform": "",
            "modulePaths": {"sap.cf":"https://portal-service.cfapps.ap10.hana.ondemand.com/resources/portal/~c23795af8ec8b47e9c2df9ea4dd94fb03c786761~/sap/cf"},
            "services": {
                "Personalization": {
                    "adapter": {
                        "module": "sap.cf.adapters.PersonalizationAdapter",
                        "config": {
                            "storageType": "REMOTE_STORAGE",
                            "remoteStorageUrl": "portal/rest/v1/personalization?siteId=default&cacheBusterToken=",
                            "isAppRuntime": true
                        }
                    }
                },
                "Container": {
                    "adapter": {
                        "config": {
                            "userProfile":{
                                "defaults": &#x7b;&quot;id&quot;&#x3a;&quot;d55e94ae-64e9-4b1c-a7a8-f50ee22797bb&quot;,&quot;email&quot;&#x3a;&quot;jason.scott&#x40;sap.com&quot;,&quot;firstName&quot;&#x3a;&quot;Jason&quot;,&quot;lastName&quot;&#x3a;&quot;Scott&quot;,&quot;fullName&quot;&#x3a;&quot;Jason&#x20;Scott&quot;,&quot;isAdminUser&quot;&#x3a;true&#x7d;
                             },
                            "systemProperties": {
                                "logoutUrl": "/cp.portal/logout",
                                "sessionKeepAlive": {
                                    "url": "/cp.portal/sessionKeepAlive",
                                    "method": "HEAD"
                                }
                            }
                        }
                    }
                }
            },
            "ui5appruntime": {
                "config": {
                    "appIndex": {
                        "module": "sap.cf.modules.tcprovider"
                    }
                }
            }
        }'>

    <script id="sap-ui-bootstrap"
            src="https://sapui5.hana.ondemand.com/1.123.1/resources/sap/fiori/appruntime-min-0.js"
            data-sap-ui-language="en"
            data-sap-ui-preload="async"
            data-sap-ui-versionedLibCss=true
            data-sap-ui-libs="sap.ui.core, sap.ushell, sap.m, sap.ui.layout"
            data-sap-ui-theme="sap_horizon"
            data-sap-ui-compatVersion="edge"
            data-sap-ui-oninit="module:sap/ushell/appRuntime/ui5/AppRuntime"
            data-sap-ui-flexibilityServices="&#x5b;&#x7b;&quot;url&quot;&#x3a;&quot;&#x2f;comsapperso&quot;,&quot;connector&quot;&#x3a;&quot;PersonalizationConnector&quot;&#x7d;&#x5d;">
    </script>

    <script src="https&#x3a;&#x2f;&#x2f;sapui5.hana.ondemand.com&#x2f;1.123.1/resources/sap/fiori/appruntime-min-1.js"></script>
    <script src="https&#x3a;&#x2f;&#x2f;sapui5.hana.ondemand.com&#x2f;1.123.1/resources/sap/fiori/appruntime-min-2.js"></script>
    <script src="https&#x3a;&#x2f;&#x2f;sapui5.hana.ondemand.com&#x2f;1.123.1/resources/sap/fiori/appruntime-min-3.js"></script>

</head>
<body class="sapUiBody" id="content">
</body>
</html>
'
    at AppRuntime-dbg.js:623:30
```

**Reply by js1972** [reply]
Sorry Willem. the second deployment fixed it all - after waiting a couple of minutes the app just started working. A cosmic ray must have hit the cpu during that first deployment.   ;-)

**Reply by js1972** [reply]
How do I enable the delta calcs? I can only see the original 3 forecast options?

<img width="418" alt="image" src="https://github.com/SAP-samples/btp-resource-consumption-monitor/assets/1317161/ff20c80d-973b-4afb-8e9e-51fe82cfe351">

**Reply by willempardaens** [reply]
I also intermittently experience the Destination Configuration Update error for some reason, and a second deployment fixes it. I will need to dig into that one next as it puzzles me a bit.

**Reply by willempardaens** [reply]
The new Delta values will be mentioned throughout the application in various places in relation to the service/metric. See for example the `Delta Cost` column in the initial Services ListView, or the `Daily Change` block in a Service ObjectPage header section.

They are calculated as the difference to the **previous day's values**, and expressed in amounts (e.g. `delta_measure_cost`) as well as in % change (e.g. `delta_measure_costPct`). Deltas are both calculated for the change in Actuals and the change in Forecasted values. So, regardless of the chosen `Forecasting Configuration` there will be delta values for you to use.

In the `Alert Configuration` you now have the additional options to trigger an alert based on delta values (either of the Actual or of the Forecasted value).

Coming back to your example of the SAP Integration Suite where a 4th tenant is activated, this would be reflected in:

- `delta_measure_cost = 4k USD`
- `delta_measure_costPct = 33%`
- `delta_measure_usage = 1 tenant` (on `metric` level)
- `delta_measure_usagePct = 33%`
- `delta_forecast_cost = 4k USD` (if `excluded`)
- ...

This first attempt focusses only on changes compared to the **previous day**, not on changes compared to the **previous month**.

Example:
<img width="758" alt="image" src="https://github.com/SAP-samples/btp-resource-consumption-monitor/assets/19388646/aea7291f-67ae-4da4-85b1-24f12f737371">

Let me know what you think.

**Comment by willempardaens** [comment]
Delivered by #11
